src/memory.jl

```diff
@@ -1,203 +1,318 @@
 # This file is a part of JuliaFEM.
 # License is MIT: see https://github.com/JuliaFEM/Materials.jl/blob/master/LICENSE
 
-@with_kw mutable struct MemoryDriverState <: AbstractMaterialState
-    time :: Float64 = zero(Float64)
-    strain :: SymmetricTensor{2,3} = zero(SymmetricTensor{2,3,Float64})
+module MemoryModule
+
+using LinearAlgebra, ForwardDiff, Tensors, NLsolve, Parameters
+
+import ..AbstractMaterial, ..AbstractMaterialState
+import ..Utilities: Symm2, Symm4, isotropic_elasticity_tensor, lame, debang
+import ..integrate_material!  # for method extension
+
+# parametrically polymorphic for any type representing ℝ
+export GenericMemory, GenericMemoryDriverState, GenericMemoryParameterState, GenericMemoryVariableState
+
+# specialization for Float64
+export Memory, MemoryDriverState, MemoryParameterState, MemoryVariableState
+
+@with_kw mutable struct GenericMemoryDriverState{T <: Real} <: AbstractMaterialState
+    time::T = zero(T)
+    strain::Symm2{T} = zero(Symm2{T})
+end
+
+# TODO: complete this docstring
+"""Parameter state for Memory material.
+
+`E`: Young's modulus
+`nu`: Poisson's ratio
+`R0`: initial yield strength
+`Kn`: plasticity multiplier divisor (drag stress)
+`nn`: plasticity multiplier exponent
+`C1`, `D1`: parameters governing behavior of backstress X1
+`C2`, `D2`: parameters governing behavior of backstress X2
+`Q0`: ???
+`QM`: ???
+`mu`: ???
+`b`: ???
+`eta`: ???
```

Comment:
Parameter that controls the balance between memory surface kinematic and isotropic hardening. Support [0,1]. With the value of 0, the memory surface hardens kinematically; with the value of 1, the memory surface hardens isotropically. Initially values of 1/2 were used by several authors. Later values <1/2 have been suggested to capture the progressive process of memorization.